### PR TITLE
Added ADR and implementation containing multiple configuration API related changes

### PR DESCRIPTION
**Pull request by gavlyukovskiy**

### CODE REVIEW

File: adr/0002-masking-configuration-API.md [hunk 1]
```diff
@@ -0,0 +1,199 @@
...
+By default we
+1. We will not make a masking decision based on type. 
+    The main reason is that masking is explicitly requested by the user by specifying the target keys (either block
+    or allow) and it would be unexpected that we do not mask an explicitly requested key because it's a number and 
+    number masking is not enabled by default.
+2. We will enable length obfuscation by default as length on its own might leak information about the original value.
+3. We will mask all types into strings.
+    This is done to keep masking consistent across types and to make it apparent that the value was masked (while `"***"`
+    is a good indicator for strings, the `888` or `0` for numbers, or `false` for booleans is not)
+4. We will use different masks for each original type in to keep information about the original type, the strings will
```
Comment (Breus) [inline adr/0002-masking-configuration-API.md:33]: Nit: remove `in` in ` type in to keep`

File: adr/0002-masking-configuration-API.md [hunk 2]
```diff
@@ -0,0 +1,199 @@
...
+    number masking is not enabled by default.
+2. We will enable length obfuscation by default as length on its own might leak information about the original value.
+3. We will mask all types into strings.
+    This is done to keep masking consistent across types and to make it apparent that the value was masked (while `"***"`
+    is a good indicator for strings, the `888` or `0` for numbers, or `false` for booleans is not)
+4. We will use different masks for each original type in to keep information about the original type, the strings will 
+    be masked as `"***"`, numbers as `"###"` and booleans as `"&&&"`.  
+    While this "leaks" additional information, that has only weak relation to original _data_, but more about 
+    implementation details and is more likely to be inferred by the key itself or be publically available.
+5. The `null` type (value) will remain unmasked until someone requests it, providing a good reason.
```
Comment (Breus) [inline adr/0002-masking-configuration-API.md:37]: Perhaps add that in this case it might be added as an **optional** feature (it won't affect the existing API)

File: src/main/java/dev/blaauwendraad/masker/json/ByteTrie.java [hunk 3]
```diff
@@ -28,21 +33,23 @@
  * <p> And lastly, we can make a small optimization to remember all the distinct lengths of the target keys, so that
  * we can fail fast if the incoming key is not of the same length.
  */
+@ParametersAreNonnullByDefault
```
Comment (Breus) [inline src/main/java/dev/blaauwendraad/masker/json/ByteTrie.java:36]: Nit: we do this in all packages now (once https://github.com/Breus/json-masker/pull/50 merges)

File: src/main/java/dev/blaauwendraad/masker/json/ByteTrie.java [hunk 4]
```diff
@@ -28,21 +33,23 @@
...
+    public ByteTrie(JsonMaskingConfig config) {
+        this.config = config;
         this.root = new TrieNode();
     }
 
     /**
      * Inserts a word into the trie.
      */
-    public void insert(String word) {
+    public void insert(String word, boolean negativeMatch) {
```
Comment (Breus) [inline src/main/java/dev/blaauwendraad/masker/json/ByteTrie.java:51]: Can you add parameters to the JavaDoc? Especially to figure out what `negativeMatch` is I had to read the method implementation

File: src/main/java/dev/blaauwendraad/masker/json/util/ValueMaskingUtil.java [hunk 5]
```diff
@@ -2,6 +2,7 @@
 
 import dev.blaauwendraad.masker.json.MaskingState;
 
+import java.nio.charset.StandardCharsets;
```
Comment (Breus) [inline src/main/java/dev/blaauwendraad/masker/json/util/ValueMaskingUtil.java:5]: I don't see this being used? Am I blind?

File: adr/0002-masking-configuration-API.md (hunk 1 above)
Comment (gavlyukovskiy) [reply adr/0002-masking-configuration-API.md:33]:
```suggestion
4. We will use different masks for each original type to keep information about the original type, the strings will 
```
FYI GitHub has suggestions feature (icon of file with +-)
image

File: adr/0002-masking-configuration-API.md [hunk 6]
```diff
@@ -0,0 +1,199 @@
...
+    or allow) and it would be unexpected that we do not mask an explicitly requested key because it's a number and 
+    number masking is not enabled by default.
+2. We will enable length obfuscation by default as length on its own might leak information about the original value.
+3. We will mask all types into strings.
+    This is done to keep masking consistent across types and to make it apparent that the value was masked (while `"***"`
+    is a good indicator for strings, the `888` or `0` for numbers, or `false` for booleans is not)
+4. We will use different masks for each original type in to keep information about the original type, the strings will 
+    be masked as `"***"`, numbers as `"###"` and booleans as `"&&&"`.  
+    While this "leaks" additional information, that has only weak relation to original _data_, but more about 
+    implementation details and is more likely to be inferred by the key itself or be publically available.
```
[the remark this answers is not in the document]
Comment (gavlyukovskiy) [reply adr/0002-masking-configuration-API.md:36]:
```suggestion
    Keeping type information by default would be helpful when debugging and, while this leaks some information about original _data_, it does not leak the data itself, but rather an implementation detail about _how_ data is stored internally. That is likely to be publicly available (e.g. by providing an OpenAPI schema) or it would be possible to infer the type from the key itself.
```

File: adr/0002-masking-configuration-API.md (hunk 2 above)
Comment (gavlyukovskiy) [reply adr/0002-masking-configuration-API.md:37]:
```suggestion
5. The `null` type (value) will remain unmasked, but we aim to make the API extendable to be able to support that if someone requests it, providing a good reason. 
```

File: src/main/java/dev/blaauwendraad/masker/json/ByteTrie.java (hunk 3 above)
Comment (gavlyukovskiy) [reply src/main/java/dev/blaauwendraad/masker/json/ByteTrie.java]: Done

File: src/main/java/dev/blaauwendraad/masker/json/ByteTrie.java (hunk 4 above)
Comment (gavlyukovskiy) [reply src/main/java/dev/blaauwendraad/masker/json/ByteTrie.java:62]: Done

File: src/main/java/dev/blaauwendraad/masker/json/ByteTrie.java [hunk 7]
```diff
@@ -28,21 +33,23 @@
...
+        boolean caseInsensitive = !config.caseSensitiveTargetKeys();
         byte[] bytes = word.getBytes(StandardCharsets.UTF_8);
```
[the remark this answers is not in the document]
Comment (gavlyukovskiy) [reply src/main/java/dev/blaauwendraad/masker/json/ByteTrie.java:64]: I think it's obvious enough since we only deal with a single string -> bytes here.

File: src/main/java/dev/blaauwendraad/masker/json/ByteTrie.java [hunk 8]
```diff
@@ -67,10 +74,10 @@ public void insert(String word) {
         TrieNode node = root;
         for (int i = 0; i < bytes.length; i++) {
             byte b = bytes[i];
```
[the remark this answers is not in the document]
Comment (gavlyukovskiy) [reply src/main/java/dev/blaauwendraad/masker/json/ByteTrie.java:87]: This is just too much @Breus, what's next `i -> index`?

File: src/main/java/dev/blaauwendraad/masker/json/ByteTrie.java [hunk 9]
```diff
@@ -84,34 +91,60 @@ public void insert(String word) {
...
             node = child;
         }
+        node.keyMaskingConfig = config.getConfig(word);
         node.endOfWord = true;
+        node.negativeMatch = negativeMatch;
     }
 
     /**
-     * Returns if the word is in the trie.
+     * If the key matches - returns its masking configuration, otherwise returns null.
```
[the remark this answers is not in the document]
Comment (gavlyukovskiy) [reply src/main/java/dev/blaauwendraad/masker/json/ByteTrie.java:117]:
```suggestion
     * If the key matches returns its masking configuration, otherwise returns null.
```

File: src/main/java/dev/blaauwendraad/masker/json/ByteTrie.java [hunk 10]
```diff
@@ -84,34 +91,60 @@ public void insert(String word) {
...
     public boolean search(byte[] bytes, int offset, int length) {
+        TrieNode node = searchNode(bytes, offset, length);
+        return node != null && !node.negativeMatch;
+    }
+
+    /**
+     * Must be called only if the key has matched or in allow mode, for key that did not match
+     * returns its masking configuration, otherwise returns default config.
+     */
+    public KeyMaskingConfig config(byte[] bytes, int offset, int length) {
```
[the remark this answers is not in the document]
Comment (gavlyukovskiy) [reply src/main/java/dev/blaauwendraad/masker/json/ByteTrie.java:128]: Let me get back to this some time after, I'm still not sure whether I want to do separate `search`/`config`. Ideally it should be one method

File: src/main/java/dev/blaauwendraad/masker/json/config/KeyMaskingConfig.java [hunk 11]
```diff
@@ -0,0 +1,377 @@
+package dev.blaauwendraad.masker.json.config;
+
+import javax.annotation.CheckForNull;
+import java.nio.charset.StandardCharsets;
+import java.util.Objects;
+
+public class KeyMaskingConfig {
```
Comment (gavlyukovskiy) [inline src/main/java/dev/blaauwendraad/masker/json/config/KeyMaskingConfig.java:7]: 🤔 
image

Comment (gavlyukovskiy) [reply src/main/java/dev/blaauwendraad/masker/json/config/KeyMaskingConfig.java:7]: But yeah, both of them can be final 😃

File: src/main/java/dev/blaauwendraad/masker/json/config/KeyMaskingConfig.java [hunk 12]
```diff
@@ -0,0 +1,368 @@
...
+
+        // Boolean masking, mutually exclusive options
+        private Boolean disableBooleanMasking;
+        private String maskBooleansWithString;
+        private Boolean maskBooleansWith;
+
+        /**
+         * Mask all string values with the provided value.
+         * For example, "maskMe": "secret" -> "maskMe": "***".
+         * <p>
```
[the remark this answers is not in the document]
Comment (gavlyukovskiy) [reply src/main/java/dev/blaauwendraad/masker/json/config/KeyMaskingConfig.java:164]: The only thing I'm not sure about, should I use `{@code }` or {@literal }` to wrap it? I had to do that for boolean's `&&&`, but now it looks slightly different 🤷

File: src/main/java/dev/blaauwendraad/masker/json/config/KeyMaskingConfig.java [hunk 13]
```diff
@@ -0,0 +1,368 @@
...
+                    maskNumbersWithString,
+                    maskNumbersWith,
+                    maskNumberDigitsWith,
+                    disableBooleanMasking != null && disableBooleanMasking,
+                    maskBooleansWithString,
+                    maskBooleansWith);
+        }
+
+        private void checkMutuallyExclusiveStringMaskingOptions() {
+            if (maskStringsWith != null || maskStringCharactersWith != null) {
```
[the remark this answers is not in the document]
Comment (gavlyukovskiy) [reply src/main/java/dev/blaauwendraad/masker/json/config/KeyMaskingConfig.java:360]: I can. Every option also has a check that it cannot be set twice:
```java
builder.maskStringsWith("***").maskStringsWith("***")
```
before this check, which means that at this time both values have to be null

File: src/main/java/dev/blaauwendraad/masker/json/config/KeyMaskingConfig.java [hunk 14]
```diff
@@ -0,0 +1,368 @@
...
+                throw new IllegalStateException("'maskStringsWith(String)' and 'maskStringCharactersWith(String)' are mutually exclusive");
+            }
+        }
+
+        private void checkMutuallyExclusiveNumberMaskingOptions() {
+            if (disableNumberMasking != null || maskNumbersWith != null || maskNumbersWithString != null || maskNumberDigitsWith != null) {
+                throw new IllegalStateException("'disableNumberMasking()', 'maskNumbersWith(int)', 'maskNumbersWith(String)' and 'maskNumberDigitsWith(int)' are mutually exclusive");
```
[the remark this answers is not in the document]
Comment (gavlyukovskiy) [reply src/main/java/dev/blaauwendraad/masker/json/config/KeyMaskingConfig.java:367]: https://github.com/Breus/json-masker/pull/47/files#r1488670642

File: src/main/java/dev/blaauwendraad/masker/json/config/KeyMaskingConfig.java [hunk 15]
```diff
@@ -0,0 +1,368 @@
...
+            }
+        }
+
+        private void checkMutuallyExclusiveBooleanMaskingOptions() {
+            if (disableBooleanMasking != null || maskBooleansWith != null || maskBooleansWithString != null) {
+                throw new IllegalStateException("'disableBooleanMasking()', 'maskBooleansWith(boolean)' and 'maskBooleansWith(String)' are mutually exclusive");
+            }
```
[the remark this answers is not in the document]
Comment (gavlyukovskiy) [reply src/main/java/dev/blaauwendraad/masker/json/config/KeyMaskingConfig.java:374]: https://github.com/Breus/json-masker/pull/47/files#r1488670642

File: src/main/java/dev/blaauwendraad/masker/json/MaskingState.java [hunk 16]
```diff
@@ -88,17 +88,16 @@ public String toString() {
      *
      * @param startIndex index from which to start replacing
      * @param endIndex   index at which to stop replacing
-     * @param maskLength length of the mask to apply
-     * @param maskByte   byte to use for the mask
+     * @param maskBytes  byte array mask to use as replacement for the value
      */
-    public record ReplacementOperation(int startIndex, int endIndex, int maskLength, byte maskByte) {
+    public record ReplacementOperation(int startIndex, int endIndex, byte[] maskBytes) {
```
[the remark this answers is not in the document]
Comment (gavlyukovskiy) [reply src/main/java/dev/blaauwendraad/masker/json/MaskingState.java]: Haha, I replaced it with `mask` before seeing your comment 😄

File: adr/0002-masking-configuration-API.md [hunk 17]
```diff
@@ -0,0 +1,199 @@
...
+    private final Set<String> targetKeys;
+    private final TargetKeyMode targetKeyMode;
+    private final MaskingConfig defaultConfig;
+    private final Map<String, MaskingConfig> targetKeyConfigs;
+    
+    public static class Builder {
+        ...
+
+        // mask mode
+        public Builder mask(String... targets) { ... };
```
Comment (donavdey) [inline adr/0002-masking-configuration-API.md:102]: Should we make the type of keys collection consistent with other `mask` overloads? That is, either use `Set` or `array` everywhere.

File: adr/0002-masking-configuration-API.md [hunk 18]
```diff
@@ -0,0 +1,199 @@
...
+        public Builder mask(String target, MaskingConfig config) { ... };
+        public Builder mask(Set<String> targets, MaskingConfig config) { ... };
+        // hypothetical custom masker, most likely will be its own type instead of Function
+        public Builder mask(String target, Function<String | Number | Boolean, String> mapper) { ... };
```
Comment (donavdey) [inline adr/0002-masking-configuration-API.md:106]: Such custom strategies will be nice to have. The type of the function is a bit tricky point though. If we are going to apply the function to the found key, then the type should be `byte[] -> byte[]` for performance considerations as to avoid transforming byte arrays back and forth to strings. However, such typing might come accross as complex and not user friendly.

File: src/jmh/java/dev/blaauwendraad/masker/json/JsonMaskerBenchmark.java [hunk 19]
```diff
@@ -1,27 +1,12 @@
...
-import org.openjdk.jmh.annotations.BenchmarkMode;
-import org.openjdk.jmh.annotations.Fork;
-import org.openjdk.jmh.annotations.Measurement;
-import org.openjdk.jmh.annotations.Mode;
-import org.openjdk.jmh.annotations.OutputTimeUnit;
-import org.openjdk.jmh.annotations.Param;
-import org.openjdk.jmh.annotations.Scope;
-import org.openjdk.jmh.annotations.Setup;
-import org.openjdk.jmh.annotations.Warmup;
+import org.openjdk.jmh.annotations.*;
```
Comment (donavdey) [inline src/jmh/java/dev/blaauwendraad/masker/json/JsonMaskerBenchmark.java:4]: I am curious about your opinion on wildcard imports.

File: src/main/java/dev/blaauwendraad/masker/json/ByteTrie.java [hunk 20]
```diff
@@ -85,67 +102,108 @@ public void insert(String word) {
...
                      */
-                    node.children[lowerBytes[i] + MAX_BYTE_SIZE] = child;
-                    node.children[upperBytes[i] + MAX_BYTE_SIZE] = child;
+                    node.children[lowerBytes[i] + BYTE_OFFSET] = child;
+                    node.children[upperBytes[i] + BYTE_OFFSET] = child;
                 }
             }
             node = child;
         }
+        node.keyMaskingConfig = config.getConfig(word);
```
Comment (donavdey) [inline src/main/java/dev/blaauwendraad/masker/json/ByteTrie.java:111]: Nice, we can use it to check if the key should be interpreted as a jsonpath to fix the bug you mentioned yesterday.

File: src/main/java/dev/blaauwendraad/masker/json/ByteTrie.java (hunk 10 above)
[the remark this answers is not in the document]
Comment (donavdey) [reply src/main/java/dev/blaauwendraad/masker/json/ByteTrie.java:128]: I am in favor of having one method that returns a TrieNode. Then the client can decide wether it wants to access a config or check for the match. This will also remove the binding of the allow mode logic to the trie.

File: src/main/java/dev/blaauwendraad/masker/json/config/KeyMaskingConfig.java [hunk 21]
```diff
@@ -0,0 +1,377 @@
...
+    private final boolean disableNumberMasking;
+    @CheckForNull
+    private final byte[] maskNumbersWith;
+    @CheckForNull
+    private final byte[] maskNumberDigitsWith;
+    private final boolean disableBooleanMasking;
+    @CheckForNull
+    private final byte[] maskBooleansWith;
+
+    KeyMaskingConfig(
```
Comment (donavdey) [inline src/main/java/dev/blaauwendraad/masker/json/config/KeyMaskingConfig.java:21]: Should it also accept an `jsonPathKey` option on a key configuration level? This way we can solve the bug you mentioned yesterday by the solution you implemented of having a key masking config in a TrieNode.
Also, this way we won't have to make assumptions of wether the key is json path or not based on wether it starts with "$." or "$[" sequences.

File: src/main/java/dev/blaauwendraad/masker/json/config/KeyMaskingConfig.java [hunk 22]
```diff
@@ -0,0 +1,377 @@
...
+                    maskNumbersWith,
+                    maskNumberDigitsWith,
+                    disableBooleanMasking != null && disableBooleanMasking,
+                    maskBooleansWithString,
+                    maskBooleansWith);
+        }
+
+        private void checkMutuallyExclusiveStringMaskingOptions() {
+            if (maskStringsWith != null || maskStringCharactersWith != null) {
+                throw new IllegalStateException("'maskStringsWith(String)' and 'maskStringCharactersWith(String)' are mutually exclusive");
```
Comment (donavdey) [inline src/main/java/dev/blaauwendraad/masker/json/config/KeyMaskingConfig.java:361]: Should it be an IllegalArgumentException as from the client perspective they supplied the invalid key masking config for the key definition? Or is it conventional in builders to throw an illegal state on misconfiguration?

File: src/main/java/dev/blaauwendraad/masker/json/path/JsonPathParser.java [hunk 23]
```diff
@@ -22,17 +21,16 @@
...
+     * Parses an input literal into a {@link dev.blaauwendraad.masker.json.path.JsonPath} object.
+     * Throws {@link java.lang.IllegalArgumentException} when the input literal does not follow the jsonpath specification.
      *
      * @param literal a jsonpath literal to be parsed.
      * @return a {@link dev.blaauwendraad.masker.json.path.JsonPath} object parsed from the literal.
      */
     @Nonnull
     public JsonPath parse(String literal) {
         if (!(literal.startsWith("$.") || literal.startsWith("$["))) {
-            throw new IllegalArgumentException(
```
Comment (donavdey) [inline src/main/java/dev/blaauwendraad/masker/json/path/JsonPathParser.java:34]: It is time to synchronize the code style scheme in Intellij :)

File: src/test/java/dev/blaauwendraad/masker/json/MaskingNonStandardCharactersTest.java [hunk 24]
```diff
@@ -124,4 +124,65 @@ void maskingNonStandardCharactersInAllowMode() {
...
+
+    @Test
+    void maskingWithUnicodeCharacters() {
+        JsonMasker jsonMasker = JsonMasker.getMasker(
+                JsonMaskingConfig.builder()
+                        .maskKeys("💩", k -> k.maskStringCharactersWith("💩"))
+                        .build()
+        );
+
+        assertThat(jsonMasker.mask(
```
Comment (donavdey) [inline src/test/java/dev/blaauwendraad/masker/json/MaskingNonStandardCharactersTest.java:136]: Nice json

**Review by donavdey — APPROVED** [review]

File: src/jmh/java/dev/blaauwendraad/masker/json/JsonMaskerBenchmark.java (hunk 19 above)
Comment (gavlyukovskiy) [reply src/jmh/java/dev/blaauwendraad/masker/json/JsonMaskerBenchmark.java:4]: My IDEA did it again :( I always configure it to not do that, but since I use a couple of different workstations, some of them still collapse those.

File: adr/0002-masking-configuration-API.md (hunk 18 above)
Comment (gavlyukovskiy) [reply adr/0002-masking-configuration-API.md:106]: You are correct. Those are good points, and both options are bad in their own way. I don't have a final idea here, thus, I only keep it theoretical right now, proving that API is extensible enough to allow that in the future.
One way to do it would be some functional inheritance, where internal function is `Function<byte[], byte[]>`, but we provide adapters between that and `Function<String, String>` (or other types). This way users have ability to sacrifice performance for nicer UX (arguably for most cases it doesn't matter)

Comment (gavlyukovskiy) [reply adr/0002-masking-configuration-API.md:106]: You are correct. Those are good points, and both options are bad in their own way. I don't have a final idea here, thus, I only keep it theoretical right now, proving that API is extensible enough to allow that in the future.
One way to do it would be some functional inheritance, where internal function is `Function<byte[], byte[]>`, but we provide adapters between that and `Function<String, String>` (or other types). This way users have ability to sacrifice performance for nicer UX (arguably for most cases it doesn't matter)

File: src/main/java/dev/blaauwendraad/masker/json/config/KeyMaskingConfig.java (hunk 22 above)
Comment (gavlyukovskiy) [reply src/main/java/dev/blaauwendraad/masker/json/config/KeyMaskingConfig.java:361]: Should be IllegalArgumentException

File: adr/0002-masking-configuration-API.md (hunk 1 above)
Comment (Breus) [reply adr/0002-masking-configuration-API.md:33]: Ah yes good suggestion. In Gitlab internally that doesn't work so I unlearned myself this 😋

File: src/main/java/dev/blaauwendraad/masker/json/config/KeyMaskingConfig.java (hunk 12 above)
[the remark this answers is not in the document]
Comment (Breus) [reply src/main/java/dev/blaauwendraad/masker/json/config/KeyMaskingConfig.java:164]: I think literal? This isn't really code, right?

File: src/main/java/dev/blaauwendraad/masker/json/ByteTrie.java (hunk 10 above)
[the remark this answers is not in the document]
Comment (gavlyukovskiy) [reply src/main/java/dev/blaauwendraad/masker/json/ByteTrie.java:128]: @donavdey I've solved it by returning a config instead, now that `ByteTrie` does much more than just a trie, I have renamed it to `KeyMatcher`

File: adr/0002-masking-configuration-API.md (hunk 17 above)
Comment (gavlyukovskiy) [reply adr/0002-masking-configuration-API.md:104]: @Breus any thoughts?

File: src/jmh/java/dev/blaauwendraad/masker/json/JsonMaskerBenchmark.java (hunk 19 above)
Comment (gavlyukovskiy) [reply src/jmh/java/dev/blaauwendraad/masker/json/JsonMaskerBenchmark.java:4]: Should be fixed now.

File: src/main/java/dev/blaauwendraad/masker/json/config/KeyMaskingConfig.java (hunk 21 above)
Comment (gavlyukovskiy) [reply src/main/java/dev/blaauwendraad/masker/json/config/KeyMaskingConfig.java:21]: I've been thinking about it a bit and I think it's not that important to solve that bug as it's pretty rare that it could happen. In order to fix it I would need to also split configs by whether that's a key or a json paths (will complicate a code a bit), or, as we discussed, associate config with a specific key (would make our fluent API slightly worse)

File: src/main/java/dev/blaauwendraad/masker/json/KeyMatcher.java [hunk 25]
```diff
@@ -0,0 +1,248 @@
...
+     *                      {@code
+     *                      builder.allow("name", "age").mask("ssn", k -> k.maskStringsWith("[redacted]"))
+     *                      }
+     *                      would only allow {@code name} and {@code age} to be present in the JSON, it would use
+     *                      default configuration to mask any other key, but would specifically mask {@code ssn} with
+     *                      a string "[redacted]". To make it possible to store just the masking configuration we
+     *                      insert a "negative match" node, that would not be treated as a target key, but provide
+     *                      a fast lookup for the configuration
+     */
+    @SuppressWarnings("java:S2259")
```
Comment (gavlyukovskiy) [inline src/main/java/dev/blaauwendraad/masker/json/KeyMatcher.java:68]:
```suggestion
    @SuppressWarnings("java:S2259") // sonar complains that lowerBytes can be null, but it can not
```

**Comment by gavlyukovskiy** [comment]
Run some benchmarks (same machine as the version from master).
TLDR:
- Before:
```
JsonMaskerBenchmark.jsonMaskerString        unicode         2mb                    0.01  thrpt           129,351  ops/s
```
- After:
```
JsonMaskerBenchmark.jsonMaskerString        unicode       false         2mb                     0.1  thrpt    8       96.084
```

One thing I noticed also when @donavdey was posting his results - json path is only in 1kb version, apparently due to some errors:
```
java.lang.IllegalArgumentException: Illegal jsonpath literal. Filter selectors are not supported.
        at dev.blaauwendraad.masker.json.path.JsonPathParser.validateSegment(JsonPathParser.java:92)
        at java.base/java.util.ArrayList.forEach(ArrayList.java:1511)
        at dev.blaauwendraad.masker.json.path.JsonPathParser.parse(JsonPathParser.java:42)
        at dev.blaauwendraad.masker.json.config.JsonMaskingConfig$Builder.maskJsonPaths(JsonMaskingConfig.java:215)
        at dev.blaauwendraad.masker.json.config.JsonMaskingConfig$Builder.maskJsonPaths(JsonMaskingConfig.java:198)
        at dev.blaauwendraad.masker.json.JsonMaskerBenchmark$State.setup(JsonMaskerBenchmark.java:50)
```
```
java.lang.IllegalArgumentException: Illegal jsonpath literal. Escape characters are not supported.
        at dev.blaauwendraad.masker.json.path.JsonPathParser.parse(JsonPathParser.java:36)
        at dev.blaauwendraad.masker.json.config.JsonMaskingConfig$Builder.maskJsonPaths(JsonMaskingConfig.java:215)
        at dev.blaauwendraad.masker.json.config.JsonMaskingConfig$Builder.maskJsonPaths(JsonMaskingConfig.java:198)
        at dev.blaauwendraad.masker.json.JsonMaskerBenchmark$State.setup(JsonMaskerBenchmark.java:50)
```
Seems like random generator produces keys, that are considered invalid segments.

<details>
  <summary>Full results:</summary>

```
Benchmark                              (characters)  (jsonPath)  (jsonSize)  (maskedKeyProbability)   Mode  Cnt        Score       Error  Units
BaselineBenchmark.countBytes                unicode         N/A         1kb                    0.01  thrpt    8  4312401.362 ▒ 39394.295  ops/s
BaselineBenchmark.countBytes                unicode         N/A       128kb                    0.01  thrpt    8    33878.588 ▒   487.540  ops/s
BaselineBenchmark.countBytes                unicode         N/A         2mb                    0.01  thrpt    8     2132.785 ▒   111.053  ops/s
BaselineBenchmark.jacksonParseAndMask       unicode         N/A         1kb                    0.01  thrpt    8    18759.527 ▒   310.547  ops/s
BaselineBenchmark.jacksonParseAndMask       unicode         N/A       128kb                    0.01  thrpt    8      116.723 ▒     1.644  ops/s
BaselineBenchmark.jacksonParseAndMask       unicode         N/A         2mb                    0.01  thrpt    8        5.870 ▒     0.107  ops/s
BaselineBenchmark.regexReplace              unicode         N/A         1kb                    0.01  thrpt    8    11950.993 ▒    54.069  ops/s
BaselineBenchmark.regexReplace              unicode         N/A       128kb                    0.01  thrpt    8       61.248 ▒     0.445  ops/s
BaselineBenchmark.regexReplace              unicode         N/A         2mb                    0.01  thrpt    8        2.784 ▒     0.026  ops/s
JsonMaskerBenchmark.jsonMaskerBytes           ascii       false         1kb                    0.01  thrpt    8   373688.648 ▒ 34522.295  ops/s
JsonMaskerBenchmark.jsonMaskerBytes           ascii       false         1kb                     0.1  thrpt    8   368104.184 ▒  3190.802  ops/s
JsonMaskerBenchmark.jsonMaskerBytes           ascii       false       128kb                    0.01  thrpt    8     1798.003 ▒    33.739  ops/s
JsonMaskerBenchmark.jsonMaskerBytes           ascii       false       128kb                     0.1  thrpt    8     1948.592 ▒    67.199  ops/s
JsonMaskerBenchmark.jsonMaskerBytes           ascii       false         2mb                    0.01  thrpt    8      120.374 ▒    16.565  ops/s
JsonMaskerBenchmark.jsonMaskerBytes           ascii       false         2mb                     0.1  thrpt    8      121.676 ▒    15.408  ops/s
JsonMaskerBenchmark.jsonMaskerBytes           ascii        true         1kb                    0.01  thrpt    8   342834.748 ▒ 17249.644  ops/s
JsonMaskerBenchmark.jsonMaskerBytes         unicode       false         1kb                    0.01  thrpt    8   492749.853 ▒ 71940.171  ops/s
JsonMaskerBenchmark.jsonMaskerBytes         unicode       false         1kb                     0.1  thrpt    8   403407.460 ▒  4062.588  ops/s
JsonMaskerBenchmark.jsonMaskerBytes         unicode       false       128kb                    0.01  thrpt    8     2549.499 ▒   490.279  ops/s
JsonMaskerBenchmark.jsonMaskerBytes         unicode       false       128kb                     0.1  thrpt    8     2567.302 ▒   387.063  ops/s
JsonMaskerBenchmark.jsonMaskerBytes         unicode       false         2mb                    0.01  thrpt    8      156.228 ▒     3.416  ops/s
JsonMaskerBenchmark.jsonMaskerBytes         unicode       false         2mb                     0.1  thrpt    8      155.212 ▒     1.101  ops/s
JsonMaskerBenchmark.jsonMaskerBytes         unicode        true         1kb                    0.01  thrpt    8   450059.957 ▒   735.255  ops/s
JsonMaskerBenchmark.jsonMaskerString          ascii       false         1kb                    0.01  thrpt    8   339380.842 ▒  1025.432  ops/s
JsonMaskerBenchmark.jsonMaskerString          ascii       false         1kb                     0.1  thrpt    8   363628.065 ▒  1402.311  ops/s
JsonMaskerBenchmark.jsonMaskerString          ascii       false       128kb                    0.01  thrpt    8     1996.894 ▒   264.570  ops/s
JsonMaskerBenchmark.jsonMaskerString          ascii       false       128kb                     0.1  thrpt    8     1991.907 ▒     8.991  ops/s
JsonMaskerBenchmark.jsonMaskerString          ascii       false         2mb                    0.01  thrpt    8      124.393 ▒     3.197  ops/s
JsonMaskerBenchmark.jsonMaskerString          ascii       false         2mb                     0.1  thrpt    8      126.643 ▒     1.983  ops/s
JsonMaskerBenchmark.jsonMaskerString          ascii        true         1kb                    0.01  thrpt    8   330144.249 ▒ 12042.831  ops/s
JsonMaskerBenchmark.jsonMaskerString        unicode       false         1kb                    0.01  thrpt    8   295806.894 ▒ 36086.378  ops/s
JsonMaskerBenchmark.jsonMaskerString        unicode       false         1kb                     0.1  thrpt    8   255393.155 ▒  2049.627  ops/s
JsonMaskerBenchmark.jsonMaskerString        unicode       false       128kb                    0.01  thrpt    8     1428.565 ▒     3.720  ops/s
JsonMaskerBenchmark.jsonMaskerString        unicode       false       128kb                     0.1  thrpt    8     1533.659 ▒    77.941  ops/s
JsonMaskerBenchmark.jsonMaskerString        unicode       false         2mb                    0.01  thrpt    8       94.205 ▒     1.885  ops/s
JsonMaskerBenchmark.jsonMaskerString        unicode       false         2mb                     0.1  thrpt    8       96.084 ▒     0.574  ops/s
JsonMaskerBenchmark.jsonMaskerString        unicode        true         1kb                    0.01  thrpt    8   299728.613 ▒  4442.899  ops/s
```
</details>

File: src/main/java/dev/blaauwendraad/masker/json/KeyContainsMasker.java [hunk 26]
```diff
@@ -370,72 +369,81 @@ private void maskArrayValueInPlace(MaskingState maskingState) {
...
+            // if we got a maskingConfig for the key - we need to mask this key with that config, but if the config we
+            // got was the default config, then it means that the key doesn't have a specific configuration and we
+            // should fallback to key specific config, that the object is being masked with
+            // e.g. for '{ "a": { "b": "value" } }' we want to use config of 'b' if any, but fallback to config of 'a'
+            // however if we're in the allow mode, then getting a null as config, means that the key has been explicitly
+            // allowed and must not be masked, even if enclosing object is being masked
+            boolean valueAllowed = maskingConfig.isInAllowMode() && keyMaskingConfig == null;
+            if (keyMaskingConfig == null || keyMaskingConfig == maskingConfig.getDefaultConfig()) {
+                keyMaskingConfig = parentKeyMaskingConfig;
             }
```
Comment (Breus) [inline src/main/java/dev/blaauwendraad/masker/json/KeyContainsMasker.java:406]: Move it down to the use

Comment (Breus) [reply src/main/java/dev/blaauwendraad/masker/json/KeyContainsMasker.java:406]: Also split comment between allow mode and config override

File: src/main/java/dev/blaauwendraad/masker/json/KeyContainsMasker.java [hunk 27]
```diff
@@ -514,6 +529,26 @@ private static void skipAllValues(MaskingState maskingState) {
         }
     }
 
+    private void maskBooleanValueInPlace(MaskingState maskingState, KeyMaskingConfig keyMaskingConfig) {
+        int targetValueLength;
+        if (AsciiCharacter.isLowercaseT(maskingState.byteAtCurrentIndex())) {
+            targetValueLength = 4;
+        } else {
+            targetValueLength = 5;
+        }
```
Comment (Breus) [inline src/main/java/dev/blaauwendraad/masker/json/KeyContainsMasker.java:538]: Turnary?

Comment (Breus) [reply src/main/java/dev/blaauwendraad/masker/json/KeyContainsMasker.java:538]:
```suggestion
        int targetValueLength = AsciiCharacter.isLowercaseT(maskingState.byteAtCurrentIndex()) ? 4 : 5;
```
